File: src/staircase.py
import numpy as np
import autograd.numpy as anp

import pymanopt
from pymanopt.manifolds import Product, Stiefel
from pymanopt.manifolds import SpecialOrthogonalGroup
from pymanopt.manifolds import Oblique
from pymanopt import Problem
from pymanopt.optimizers import TrustRegions

np.random.seed(1)


def solve_problem(Q, d, r0, dn, n):
    dn = min(dn, d)  # Ensure dn does not exceed d

    #for r in range(r0, dn + 1):
    #for r in range(2, d+1):
    for r in range(3, 4):
        print('r: ', r)
        manifold = Product([Stiefel(d, r) for _ in range(n)])
        #manifold = Product([Oblique(d, r) for _ in range(n)])
        #manifold = Product([SpecialOrthogonalGroup(d) for _ in range(n)])

        # Define the cost function using the appropriate decorator
        @pymanopt.function.numpy(manifold)
        def cost(*X):
            X_stacked = np.vstack(X)
            return anp.trace(anp.dot(X_stacked.T, anp.dot(Q, X_stacked)))

        # Define the gradient of the cost function using the appropriate decorator
        @pymanopt.function.numpy(manifold)
        def egrad(*X):
            X_stacked = np.vstack(X)
            grad_stacked = 2 * anp.dot(Q, X_stacked)
            return [grad_stacked[i * d:(i + 1) * d] for i in range(n)]

        # Define the Hessian of the cost function using the appropriate decorator
        @pymanopt.function.autograd(manifold)
        def ehess(*X_and_U):
            n = len(X_and_U) // 2
            X = X_and_U[:n]
            U = X_and_U[n:]
            U_stacked = anp.vstack(U)
            hess_stacked = 2 * anp.dot(Q, U_stacked)
            return [hess_stacked[i * d:(i + 1) * d] for i in range(n)]

        problem = Problem(manifold=manifold, cost=cost, euclidean_gradient=egrad, euclidean_hessian=ehess)
        solver = TrustRegions(min_gradient_norm=1e-4)
        Y_star = solver.run(problem).point

        if any(np.linalg.matrix_rank(y) < r for y in Y_star):
            return np.vstack(Y_star)
        else:
            if r < dn:
                Y_star = [np.pad(y, ((0, 0), (0, d))) for y in Y_star]

    return np.vstack(Y_star)

File: src/staircase.m
function Y_star = solve_problem(Q, n)

    d = 3
    dn = d * n
    r0 = 6
    rmax = 10

    % Initialize the random seed for reproducibility
    rng(1);

    % Compute the Jacobi preconditioner (inverse of the diagonal elements of Q)
    P = 1 ./ diag(Q);

    for r = r0:rmax
        disp(['r: ', num2str(r)]);
        
        % Define the manifold using Manopt
        manifold = stiefelstackedfactory(n, d, r);

        % Define the cost function
        cost = @(X) cost_function(X, Q);

        % Define the gradient of the cost function
        egrad = @(X) egrad_function(X, Q, P);

        % Define the Hessian of the cost function
        ehess = @(X, U) ehess_function(X, U, Q, P);

        % Set up the problem structure
        problem.M = manifold;
        problem.cost = cost;
        problem.egrad = egrad;
        problem.ehess = ehess;

        % Solve the problem using the trust-regions solver
        options.tolgradnorm = 1e-2;
        options.rel_func_tol = 1e-5;
        options.miniter = 1;
        options.maxiter = 300;
        options.maxinner = 500;
        [Y_star, ~] = trustregions(problem, [], options);

        % Check rank condition and return if satisfied
        Y_star = mat2cell(Y_star, d*ones(n, 1), r);
        if any(cellfun(@(y) rank(y) < r, Y_star))
            Y_star = cell2mat(Y_star);
            return;
        else
            if r < dn
                Y_star = cellfun(@(y) padarray(y, [0, d - size(y, 2)], 0, 'post'), Y_star, 'UniformOutput', false);
            end
        end
    end

    Y_star = cell2mat(Y_star);
end

function f = cost_function(X, Q)
    f = trace(X' * Q * X);
end

function g = egrad_function(X, Q, P)
    g = 2 * Q * X;
    %g = bsxfun(@times, P, g); % Apply Jacobi preconditioner
end

function h = ehess_function(X, U, Q, P)
    h = 2 * Q * U;
    %h = bsxfun(@times, P, h); % Apply Jacobi preconditioner
end

